sdks/python/apache_beam/runners/interactive/interactive_runner.py

```diff
@@ -76,27 +75,47 @@ def run_pipeline(self, pipeline):
 
     pcolls_to_pcoll_id = self._pcolls_to_pcoll_id(pipeline, original_context)
 
+    # Make a copy of the original pipeline to avoid accidental manipulation
+    pipeline, context = beam.pipeline.Pipeline.from_runner_api(
+        pipeline_proto,
+        self._underlying_runner,
+        pipeline._options,  # pylint: disable=protected-access
+        return_context=True)
     pipeline_info = PipelineInfo(pipeline_proto.components)
-    context = pipeline_context.PipelineContext(pipeline_proto.components)
 
     caches_used = set()
 
     def _producing_transforms(pcoll_id, leaf=False):
       """Returns PTransforms (and their names) that produces the given PColl."""
       derivation = pipeline_info.derivation(pcoll_id)
       if self._cache_manager.exists('full', derivation.cache_label()):
+        # If the PCollection is cached, yield ReadCache PTransform that reads
+        # the PCollection and all its sub PTransforms.
         if not leaf:
           caches_used.add(pcoll_id)
-          yield ('Read' + derivation.cache_label(),
-                 beam_runner_api_pb2.PTransform(
-                     unique_name='Read' + derivation.cache_label(),
-                     spec=beam.io.Read(
-                         beam.io.ReadFromText(
-                             self._cache_manager.glob_path(
-                                 'full', derivation.cache_label()),
-                             coder=SafeFastPrimitivesCoder())._source)
-                     .to_runner_api(context),
-                     outputs={'None': pcoll_id}))
+
+          cache_label = pipeline_info.derivation(pcoll_id).cache_label()
+          dummy_pcoll = pipeline | ReadCache(self._cache_manager, cache_label)
+
+          # Find the top level ReadCache composite PTransform.
+          read_cache = dummy_pcoll.producer
+          while read_cache.parent.parent:
+            read_cache = read_cache.parent
```

Comment:
Can we make this less fragile?  This will not work, depending on how many levels deep the composite is nested.  We can make the termination condition that `isinstance(read_cache.transform, ...)`.

Reply:
I don't think the level should affect? ReadCache (and all other PTransforms applied directly to the pipeline) will always be a top level PTransform, and `top_level_ptransform.parent` will always be the root transform, and `top_level_ptransform.parent.parent` will be null.

Reply:
Ah, thank you.  I missed the fact that you directly create the input `dummy_pcoll` here.